File: app/src/test/scala/tripaint/control/ActionsTest.scala
package tripaint.control

import tripaint.ScalaFxExt.given
import tripaint.TriPaintModel
import tripaint.coords.{GridCoords, StorageCoords}
import tripaint.image.{ImageStorage, RegularImage}
import tripaint.image.ImagePool.SaveLocation
import tripaint.image.format.{SimpleStorageFormat, StorageFormat}
import tripaint.infrastructure.FileSystem
import tripaint.view.FileOpenSettings

import munit.FunSuite
import scalafx.scene.paint.Color as FXColor

import java.io.File
import scala.language.implicitConversions

class ActionsTest extends FunSuite {
  val storageFormat: StorageFormat = SimpleStorageFormat

  test("Actions.openImage should do nothing if loading failed") {
    val location = SaveLocation(null)
    val imageSize = 16

    val model = TriPaintModel.createNull(imageSize, FileSystem.NullArgs(initialImages = Map.empty))
    val pool = model.imagePool

    Actions.openImage(
      model,
      location.file,
      FileOpenSettings(location.offset, storageFormat),
      GridCoords(0, 0)
    )

    assertEquals(pool.imageAt(location), None)
  }

  test("Actions.openImage should store the loaded image in the image pool") {
    val file = new File("path/to/image.png")
    val location = SaveLocation(file)
    val imageSize = 16

    val image = ImageStorage.fill(imageSize, FXColor.Orange)
    val regularImage = image.toRegularImage(storageFormat)

    val model =
      TriPaintModel.createNull(
        imageSize,
        FileSystem.NullArgs(initialImages = Map(file -> regularImage))
      )

    Actions.openImage(
      model,
      location.file,
      FileOpenSettings(location.offset, storageFormat),
      GridCoords(0, 0)
    )
    val loadedImage = model.imagePool.imageAt(location).get

    assertEquals(loadedImage.toRegularImage(storageFormat), regularImage)
  }

  test("Actions.openImage should load image with offset") {
    val file = new File("path/to/image.png")
    val offset = StorageCoords(2, 3)
    val location = SaveLocation(file, offset)
    val imageSize = 16

    val image = ImageStorage.fill(imageSize, FXColor.Orange)
    val regularImage = image.toRegularImage(storageFormat)

    val storedImage = RegularImage.ofSize(imageSize + offset.x, imageSize + offset.y)
    storedImage.pasteImage(offset, regularImage)

    val model = TriPaintModel.createNull(
      imageSize,
      FileSystem.NullArgs(initialImages = Map(file -> storedImage))
    )
    val pool = model.imagePool

    Actions.openImage(
      model,
      location.file,
      FileOpenSettings(location.offset, storageFormat),
      GridCoords(0, 0)
    )
    val loadedImage = pool.imageAt(location).get

    assertEquals(loadedImage.toRegularImage(storageFormat), regularImage)
  }
}

File: app/src/test/kotlin/tripaint/app/ActionsTest.kt
package tripaint.app

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Nested
import tripaint.color.Color
import tripaint.coords.GridCoords
import tripaint.coords.StorageCoords
import tripaint.grid.ImageGrid
import tripaint.image.ImagePool
import tripaint.image.ImageStorage
import tripaint.image.RegularImage
import tripaint.image.format.SimpleStorageFormat
import tripaint.image.format.StorageFormat
import tripaint.view.FileOpenSettings
import java.io.File
import kotlin.test.Test

class ActionsTest {
    private val storageFormat: StorageFormat = SimpleStorageFormat

    @Nested
    inner class OpenImage {
        @Test
        fun `does nothing if loading fails`() {
            val location = ImagePool.SaveLocation(File("a.png"))
            val imageSize = 16

            val fileSystem = FileSystem.createNull(FileSystem.NullArgs(initialImages = mapOf()))
            val imagePool = ImagePool()
            val imageGrid = ImageGrid(imageSize)

            Actions.openImage(
                fileSystem, imagePool, imageGrid,
                location.file,
                FileOpenSettings(location.offset, storageFormat),
                GridCoords.from(0, 0)
            )

            assertEquals(null, imagePool.imageAt(location))
        }

        @Test
        fun `stores the loaded image in the image pool`() {
            val file = File("path/to/image.png")
            val location = ImagePool.SaveLocation(file)
            val imageSize = 16

            val image = ImageStorage.fill(imageSize, Color.Yellow)
            val regularImage = image.toRegularImage(storageFormat)

            val fileSystem = run {
                val initialImages = mapOf(Pair(file, regularImage))
                FileSystem.createNull(FileSystem.NullArgs(initialImages))
            }
            val imagePool = ImagePool()
            val imageGrid = ImageGrid(imageSize)

            Actions.openImage(
                fileSystem, imagePool, imageGrid,
                location.file,
                FileOpenSettings(location.offset, storageFormat),
                GridCoords.from(0, 0)
            )
            val loadedImage = imagePool.imageAt(location)!!

            assertEquals(regularImage, loadedImage.toRegularImage(storageFormat))
        }

        @Test
        fun `loads image with offset`() {
            val file = File("path/to/image.png")
            val offset = StorageCoords.from(2, 3)
            val location = ImagePool.SaveLocation(file, offset)
            val imageSize = 16

            val image = ImageStorage.fill(imageSize, Color.Yellow)
            val regularImage = image.toRegularImage(storageFormat)

            val storedImage = RegularImage.ofSize(imageSize + offset.x, imageSize + offset.y)
            storedImage.pasteImage(offset, regularImage)

            val fileSystem = FileSystem.createNull(FileSystem.NullArgs(initialImages = mapOf(Pair(file, storedImage))))
            val imagePool = ImagePool()
            val imageGrid = ImageGrid(imageSize)

            Actions.openImage(
                fileSystem, imagePool, imageGrid,
                location.file,
                FileOpenSettings(location.offset, storageFormat),
                GridCoords.from(0, 0)
            )
            val loadedImage = imagePool.imageAt(location)!!

            assertEquals(regularImage, loadedImage.toRegularImage(storageFormat))
        }
    }
}